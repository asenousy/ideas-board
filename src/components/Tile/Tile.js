import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { actions } from "../../store";
import styles from "./Tile.module.css";
import Idea from "../Idea/Idea";
import IdeaForm from "../IdeaForm/IdeaForm";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTrash, faEdit } from "@fortawesome/free-solid-svg-icons";
const { deleteIdea, updateIdea } = actions;

export class Tile extends PureComponent {
  constructor(props) {
    super(props);
    this.state = { editMode: props.idea.new };
  }

  render() {
    const { idea } = this.props;
    return (
      <div className={styles.wrapper} tabIndex="0">
        {this.state.editMode ? (
          <IdeaForm idea={idea} onSubmit={this.changeHandler} />
        ) : (
          <Idea
            title={idea.title}
            description={idea.description}
            date={idea.date}
          />
        )}
        {!this.state.editMode && (
          <FontAwesomeIcon
            className={styles.edit}
            icon={faEdit}
            size="lg"
            onClick={this.toggleEditMode}
            tabIndex="0"
            role="button"
            aria-label="edit"
          />
        )}
        <FontAwesomeIcon
          className={styles.delete}
          icon={faTrash}
          size="lg"
          onClick={this.deleteHandler}
          tabIndex="0"
          role="button"
          aria-label="delete"
        />
      </div>
    );
  }

  toggleEditMode = () => {
    this.setState(state => ({ editMode: !state.editMode }));
  };

  changeHandler = updatedIdea => {
    this.toggleEditMode();
    this.props.updateIdea && this.props.updateIdea(updatedIdea);
  };

  deleteHandler = () => {
    this.props.deleteIdea && this.props.deleteIdea(this.props.idea);
  };
}

Tile.propTypes = {
  idea: PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string,
    description: PropTypes.string,
    date: PropTypes.number.isRequired,
    new: PropTypes.bool.isRequired
  }).isRequired,
  deleteIdea: PropTypes.func,
  updateIdea: PropTypes.func
};

export default connect(
  null,
  { deleteIdea, updateIdea }
)(Tile);
