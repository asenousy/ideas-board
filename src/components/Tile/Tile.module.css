.wrapper {
  position: relative;
  border-radius: 5px;
  padding: 10px;
  margin: 10px;
  background-color: white;
  max-width: 200px;
}

.edit {
  position: absolute;
  top: -10px;
  right: -10px;
  padding: 10px;
  border-radius: 50px;
  color: white;
  background-color: #868e9d;
}

.delete {
  position: absolute;
  bottom: -10px;
  right: -10px;
  padding: 10px;
  border-radius: 50px;
  color: white;
  background-color: red;
}
