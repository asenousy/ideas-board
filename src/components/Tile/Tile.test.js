import React from "react";
import { shallow, mount } from "enzyme";
import { Tile } from "./Tile";

const idea = {
  id: "1",
  title: "title",
  description: "description",
  date: 111,
  new: false
};

it("renders without crashing", () => {
  const wrapper = shallow(<Tile idea={idea} />);
  expect(wrapper).toMatchSnapshot();
});

it("it should toggle to edit mode and show form", () => {
  const wrapper = mount(<Tile idea={idea} />);
  const edit = wrapper.find("FontAwesomeIcon").at(0);
  expect(wrapper.find("form").exists()).toBe(false);
  edit.simulate("click");
  expect(wrapper.state("editMode")).toBe(true);
  expect(wrapper.find("form").exists()).toBe(true);
});
