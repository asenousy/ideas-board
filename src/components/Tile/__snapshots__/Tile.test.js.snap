// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Tile
    idea={
      Object {
        "date": 111,
        "description": "description",
        "id": "1",
        "new": false,
        "title": "title",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <UNDEFINED
          date={111}
          description="description"
          title="title"
        />,
        <FontAwesomeIcon
          aria-label="edit"
          border={false}
          className="edit"
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                576,
                512,
                Array [],
                "f044",
                "M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z",
              ],
              "iconName": "edit",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          onClick={[Function]}
          onKeyUp={[Function]}
          pull={null}
          pulse={false}
          role="button"
          rotation={null}
          size="lg"
          spin={false}
          symbol={false}
          tabIndex="0"
          title=""
          transform={null}
        />,
        <FontAwesomeIcon
          aria-label="delete"
          border={false}
          className="delete"
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                448,
                512,
                Array [],
                "f1f8",
                "M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z",
              ],
              "iconName": "trash",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          onClick={[Function]}
          onKeyUp={[Function]}
          pull={null}
          pulse={false}
          role="button"
          rotation={null}
          size="lg"
          spin={false}
          symbol={false}
          tabIndex="0"
          title=""
          transform={null}
        />,
      ],
      "className": "wrapper",
      "tabIndex": "0",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "date": 111,
          "description": "description",
          "title": "title",
        },
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.memo),
          "compare": null,
          "type": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "aria-label": "edit",
          "border": false,
          "className": "edit",
          "fixedWidth": false,
          "flip": null,
          "icon": Object {
            "icon": Array [
              576,
              512,
              Array [],
              "f044",
              "M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z",
            ],
            "iconName": "edit",
            "prefix": "fas",
          },
          "inverse": false,
          "listItem": false,
          "mask": null,
          "onClick": [Function],
          "onKeyUp": [Function],
          "pull": null,
          "pulse": false,
          "role": "button",
          "rotation": null,
          "size": "lg",
          "spin": false,
          "symbol": false,
          "tabIndex": "0",
          "title": "",
          "transform": null,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "aria-label": "delete",
          "border": false,
          "className": "delete",
          "fixedWidth": false,
          "flip": null,
          "icon": Object {
            "icon": Array [
              448,
              512,
              Array [],
              "f1f8",
              "M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z",
            ],
            "iconName": "trash",
            "prefix": "fas",
          },
          "inverse": false,
          "listItem": false,
          "mask": null,
          "onClick": [Function],
          "onKeyUp": [Function],
          "pull": null,
          "pulse": false,
          "role": "button",
          "rotation": null,
          "size": "lg",
          "spin": false,
          "symbol": false,
          "tabIndex": "0",
          "title": "",
          "transform": null,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <UNDEFINED
            date={111}
            description="description"
            title="title"
          />,
          <FontAwesomeIcon
            aria-label="edit"
            border={false}
            className="edit"
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  576,
                  512,
                  Array [],
                  "f044",
                  "M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z",
                ],
                "iconName": "edit",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            onClick={[Function]}
            onKeyUp={[Function]}
            pull={null}
            pulse={false}
            role="button"
            rotation={null}
            size="lg"
            spin={false}
            symbol={false}
            tabIndex="0"
            title=""
            transform={null}
          />,
          <FontAwesomeIcon
            aria-label="delete"
            border={false}
            className="delete"
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  448,
                  512,
                  Array [],
                  "f1f8",
                  "M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z",
                ],
                "iconName": "trash",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            onClick={[Function]}
            onKeyUp={[Function]}
            pull={null}
            pulse={false}
            role="button"
            rotation={null}
            size="lg"
            spin={false}
            symbol={false}
            tabIndex="0"
            title=""
            transform={null}
          />,
        ],
        "className": "wrapper",
        "tabIndex": "0",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "date": 111,
            "description": "description",
            "title": "title",
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.memo),
            "compare": null,
            "type": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "aria-label": "edit",
            "border": false,
            "className": "edit",
            "fixedWidth": false,
            "flip": null,
            "icon": Object {
              "icon": Array [
                576,
                512,
                Array [],
                "f044",
                "M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z",
              ],
              "iconName": "edit",
              "prefix": "fas",
            },
            "inverse": false,
            "listItem": false,
            "mask": null,
            "onClick": [Function],
            "onKeyUp": [Function],
            "pull": null,
            "pulse": false,
            "role": "button",
            "rotation": null,
            "size": "lg",
            "spin": false,
            "symbol": false,
            "tabIndex": "0",
            "title": "",
            "transform": null,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "aria-label": "delete",
            "border": false,
            "className": "delete",
            "fixedWidth": false,
            "flip": null,
            "icon": Object {
              "icon": Array [
                448,
                512,
                Array [],
                "f1f8",
                "M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z",
              ],
              "iconName": "trash",
              "prefix": "fas",
            },
            "inverse": false,
            "listItem": false,
            "mask": null,
            "onClick": [Function],
            "onKeyUp": [Function],
            "pull": null,
            "pulse": false,
            "role": "button",
            "rotation": null,
            "size": "lg",
            "spin": false,
            "symbol": false,
            "tabIndex": "0",
            "title": "",
            "transform": null,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
